File: bench/bench.py
#!/usr/bin/env python3
"""quicktok vs tiktoken — single-thread and parallel batch, on bench/corpus.txt.

Both libraries are exercised through their native APIs (quicktok.encode /
encode_batch, tiktoken.encode_ordinary / encode_ordinary_batch), so this is the
apples-to-apples comparison a Python consumer actually sees. Exactness is checked
before timing. Run: make bench-py   (needs: pip install quicktok tiktoken)
"""
import os, sys, time

HERE = os.path.dirname(os.path.abspath(__file__))
CORPUS = os.path.join(HERE, "corpus.txt")


def best_of(reps, fn):
    fn()
    best = float("inf")
    for _ in range(reps):
        t0 = time.perf_counter(); fn(); best = min(best, time.perf_counter() - t0)
    return best


def main():
    try:
        import quicktok, tiktoken
    except ImportError as e:
        sys.exit(f"need quicktok + tiktoken installed: {e}")

    text = open(CORPUS, encoding="utf-8").read()
    docs = [d for d in text.split("\n\n") if d]
    mb = len(text.encode()) / 1e6
    nthreads = os.cpu_count() or 8
    print(f"quicktok {quicktok.__version__} vs tiktoken {tiktoken.__version__}")
    print(f"corpus {mb:.2f} MB, {len(docs)} docs, {nthreads} cores\n")

    for enc in ("cl100k_base", "o200k_base"):
        qt = quicktok.get_encoding(enc)
        tk = tiktoken.get_encoding(enc)
        # exactness gate
        assert qt.encode(text) == tk.encode_ordinary(text), f"{enc}: id mismatch!"
        ntok = len(qt.encode(text))
        print(f"[{enc}]  {ntok} tokens, exact vs tiktoken ✓")

        q1 = best_of(5, lambda: qt.encode(text))
        t1 = best_of(5, lambda: tk.encode_ordinary(text))
        print(f"  single-thread   quicktok {mb/q1:7.1f} MB/s   tiktoken {mb/t1:6.1f} MB/s   {t1/q1:.1f}x")

        T = nthreads
        qb = best_of(5, lambda: qt.encode_batch(docs, T))
        qn = best_of(5, lambda: qt.encode_batch_numpy(docs, T))
        tb = best_of(5, lambda: tk.encode_ordinary_batch(docs, num_threads=T))
        print(f"  batch ({T:>2} thr) list    quicktok {mb/qb:7.1f} MB/s   tiktoken {mb/tb:6.1f} MB/s   {tb/qb:.1f}x")
        print(f"  batch ({T:>2} thr) numpy   quicktok {mb/qn:7.1f} MB/s   ({q1/qn:.1f}x its single-thread, {tb/qn:.0f}x tiktoken batch)\n")


if __name__ == "__main__":
    main()

File: bench/bench.cpp
// quicktok native benchmark — single-thread throughput + parallel batch scaling.
// Self-contained: uses bench/corpus.txt (public-domain) and data/. Reports the
// numbers; no extrapolation. For cross-encoder comparisons see the main README.
//
//   make bench   (builds + runs this)
#include "quicktok.hpp"
#include <chrono>
#include <cstdio>
#include <string>
#include <thread>
#include <algorithm>
#include <vector>
using clk = std::chrono::steady_clock;

static std::string rd(const char* p) {
    FILE* f = fopen(p, "rb");
    if (!f) { fprintf(stderr, "cannot open %s\n", p); exit(1); }
    fseek(f, 0, SEEK_END); long n = ftell(f); fseek(f, 0, SEEK_SET);
    std::string s(n, 0); if (n && fread(&s[0], 1, n, f) != (size_t)n) exit(1); fclose(f); return s;
}

template <class F> static double best_of(int reps, F&& f) {
    f();  // warm
    double best = 1e30;
    for (int r = 0; r < reps; r++) { auto t0 = clk::now(); f(); best = std::min(best, std::chrono::duration<double>(clk::now() - t0).count()); }
    return best;
}

int main(int argc, char** argv) {
    std::string dir = argc > 1 ? argv[1] : "data";
    std::string corpus = argc > 2 ? argv[2] : "bench/corpus.txt";
    std::string text = rd(corpus.c_str());
    double MB = text.size() / 1e6;
    unsigned hw = std::thread::hardware_concurrency();
    printf("quicktok native bench — corpus %.2f MB, %u hardware threads\n\n", MB, hw);

    // split into documents (paragraphs) for the batch test
    std::vector<std::string> docs; std::vector<std::string_view> views;
    { size_t i = 0; while (i < text.size()) {
        size_t j = text.find("\n\n", i); if (j == std::string::npos) j = text.size();
        if (j > i) docs.emplace_back(text.substr(i, j - i)); i = j + 2; } }
    for (auto& d : docs) views.emplace_back(d);

    for (const char* enc : {"cl100k_base", "o200k_base"}) {
        auto tok = quicktok::Tokenizer::load_dir(dir, enc);
        std::vector<uint32_t> out; out.reserve(text.size() / 3);
        double st = best_of(7, [&]{ out.clear(); tok.encode((const uint8_t*)text.data(), text.size(), out); });
        size_t ntok = out.size();
        printf("[%s]  vocab %zu, %zu docs, %zu tokens\n", enc, tok.vocab_size(), docs.size(), ntok);
        printf("  single-thread : %7.1f MB/s   %6.2f Mtok/s\n", MB / st, ntok / st / 1e6);

        printf("  batch (encode_batch):\n");
        // thread ladder: powers of two, the machine's core count, and 2x cores to
        // show where it saturates / oversubscribes. Deduped + sorted.
        std::vector<unsigned> ladder = {1, 2, 4, 8, hw ? hw : 8u, hw ? 2u * hw : 16u};
        std::sort(ladder.begin(), ladder.end());
        ladder.erase(std::unique(ladder.begin(), ladder.end()), ladder.end());
        double t1 = 0;
        for (unsigned T : ladder) {
            double t = best_of(5, [&]{ auto r = tok.encode_batch(views, T); (void)r; });
            if (T == 1) t1 = t;
            printf("    %2u thread%-2s: %7.1f MB/s   %6.2f Mtok/s   %4.2fx\n",
                   T, T == 1 ? "" : "s", MB / t, ntok / t / 1e6, t1 / t);
        }
        printf("\n");
    }
    return 0;
}
